## Q1. What is the difference between Ordinal Encoding and Label Encoding? Provide an example of when you  might choose one over the other.

Ans:
    
Difference between these two techniques:

Ordinal encoding is a method of encoding categorical data where each unique category is assigned an integer value based on its order or rank. For example, if we have a categorical feature "education level" with categories "high school", "some college", "associate degree", "bachelor's degree", and "graduate degree", we could assign values of 1, 2, 3, 4, and 5 respectively, based on the order of education level.
Label encoding is a method of encoding categorical data where each unique category is assigned a unique integer value. For example, if we have a categorical feature "color" with categories "red", "blue", and "green", we could assign values of 1, 2, and 3 respectively, based on the order in which they appear in the dataset.

In general, if there is an inherent order or ranking in the categories, it makes sense to use ordinal encoding. For example, if we have a feature "income level" with categories "low", "medium", and "high", it makes sense to use ordinal encoding because there is a clear order to the categories.
On the other hand, if there is no inherent order or ranking in the categories, it makes more sense to use label encoding. For example, if we have a feature "favorite fruit" with categories "apple", "banana", "orange", and "mango", there is no clear order to the categories, so label encoding would be more appropriate.

## Q2. Explain how Target Guided Ordinal Encoding works and provide an example of when you might use it in  a machine learning project.

Ans:
    
Target Guided Ordinal Encoding is a technique used to encode categorical variables based on the target variable. This method creates a monotonic relationship between the encoded variable and the target variable, which can be helpful in improving the predictive power of the encoded variable.

Here's how Target Guided Ordinal Encoding works:

1. Group the categories of the categorical variable by their mean target value. For example, if we have a categorical variable "city" with four categories ("New York", "Los Angeles", "Chicago", "Houston") and corresponding target values of [0.3, 0.5, 0.2, 0.4], we would group the categories as follows: "Los Angeles", "Houston", "New York", "Chicago", since their corresponding mean target values are 0.5, 0.4, 0.3, and 0.2 respectively.
2. Assign an ordinal number to each group based on its mean target value. The group with the highest mean target value is assigned the highest number, and so on. In the example above, we would assign the numbers 4, 3, 2, and 1 respectively.
3. Replace the original categorical variable with the assigned ordinal numbers.

Target Guided Ordinal Encoding can be useful in a machine learning project when the categorical variable has a strong relationship with the target variable and there is a need to capture this relationship in the encoded variable.
For example, if we are building a model to predict customer churn, and we have a categorical variable "subscription type" with categories "basic", "premium", and "ultimate", we could use Target Guided Ordinal Encoding to create an encoded variable that captures the relationship between subscription type and churn. This encoded variable could then be used as a feature in our machine learning model to improve its predictive power.

## Q3. Define covariance and explain why it is important in statistical analysis. How is covariance calculated?

Ans:
    
Covariance is a measure of the linear relationship between two variables.A positive covariance means that the variables tend to increase or decrease together, while a negative covariance means that one variable tends to increase when the other variable decreases.

Covariance is an important statistical concept because it helps us to understand the relationship between two variables. For example, if we are analyzing the relationship between education level and income, we can calculate the covariance between these two variables to determine if they are positively or negatively related. If we find a positive covariance, we can conclude that people with higher education levels tend to have higher incomes, while a negative covariance would suggest the opposite.

Covariance is calculated as follows:

cov(X, Y) = E[(X - E[X])(Y - E[Y])]

where X and Y are the two variables, E[X] and E[Y] are their respective means, and E[] denotes the expected value. The formula subtracts the mean value of each variable from their respective values, multiplies the resulting differences, and takes the expected value of the product.

The resulting value can be positive, negative, or zero. A positive value indicates that the two variables tend to move in the same direction, a negative value indicates that they tend to move in opposite directions, and a value of zero indicates that there is no linear relationship between the two variables.

## Q4. For a dataset with the following categorical variables: Color (red, green, blue), Size (small, medium, large), and Material (wood, metal, plastic), perform label encoding using Python's scikit-learn library. Show your code and explain the output.

In [4]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
# Define the data as a list of lists
data = [['red', 'small', 'wood'],
        ['green', 'medium', 'metal'],
        ['blue', 'large', 'plastic'],
        ['red', 'small', 'plastic']]
# Define the column names
columns = ['Color', 'Size', 'Material']
# Create a DataFrame
df = pd.DataFrame(data, columns=columns)
# Print Dataframe before encoding
print(f'Dataframe Before Encoding :\n {df}')
print('\n')
# Create a LabelEncoder object
le = LabelEncoder()
# Apply label encoding to each column in the DataFrame
for col in df.columns:
    df[col] = le.fit_transform(df[col])
# Print the encoded DataFrame
print(f'Dataframe After Encoding :\n {df}')

Dataframe Before Encoding :
    Color    Size Material
0    red   small     wood
1  green  medium    metal
2   blue   large  plastic
3    red   small  plastic


Dataframe After Encoding :
    Color  Size  Material
0      2     2         2
1      1     1         0
2      0     0         1
3      2     2         1


In the encoded dataset, each categorical variable has been replaced with numerical values. For example, 'red' is encoded as 2, 'green' as 1, and 'blue' as 0 for the 'Color' variable. Similarly, 'small' is encoded as 2, 'medium' as 0, and 'large' as 1 for the 'Size' variable, and 'wood' is encoded as 2, 'metal' as 1, and 'plastic' as 0 for the 'Material' variable.

This encoding is done based on alphabetical order eg. blue = 0 , green = 1 , red = 2.

## Q5. Calculate the covariance matrix for the following variables in a dataset: Age, Income, and Education  level. Interpret the results.

Ans:

To calculate the covariance matrix for the variables Age, Income, and Education level, we can use the following formula: 
    
    cov(Age, Age)    cov(Age, Income)    cov(Age, Education)
    cov(Income, Age)  cov(Income, Income)  cov(Income, Education)
    cov(Education, Age) cov(Education, Income) cov(Education, Education)
Where cov(x,y) represents the covariance between variable x and y.

In [16]:
# example in python
    
import numpy as np
import pandas as pd
# create a sample dataset
data = pd.DataFrame({
    'Age': [32, 45, 28, 36, 47],
    'Income': [60000, 75000, 50000, 65000, 80000],
    'Education': [14, 18, 12, 15, 20]
})
print('Dataframe: \n',data,'\n')
# calculate the covariance matrix
cov_matrix = np.cov(data.T)
# print the covariance matrix
print("Covariance matrix:\n",cov_matrix)


Dataframe: 
    Age  Income  Education
0   32   60000         14
1   45   75000         18
2   28   50000         12
3   36   65000         15
4   47   80000         20 

Covariance matrix:
 [[6.730e+01 9.675e+04 2.590e+01]
 [9.675e+04 1.425e+08 3.775e+04]
 [2.590e+01 3.775e+04 1.020e+01]]


The resulting covariance matrix shows the covariance values between each pair of variables. For example, the covariance between Age and Income is positive, indicating that as Age increases, Income tends to increase as well.The diagonal elements in the matrix represent the variance of each variable.

## Q6. You are working on a machine learning project with a dataset containing several categorical  variables, including "Gender" (Male/Female), "Education Level" (High School/Bachelor's/Master's/PhD),  and "Employment Status" (Unemployed/Part-Time/Full-Time). Which encoding method would you use for each variable, and why?

# Ans:

For the given categorical variables "Gender", "Education Level", and "Employment Status", we can use different encoding methods based on the nature of the data and the machine learning algorithm we plan to use. Here are some possible encoding methods:
1. Gender: Since there are only two categories (Male/Female), we can use binary encoding. We can assign 0 to Male and 1 to Female. Alternatively, we can use label encoding, where we assign 0 to one category and 1 to the other category. The choice between binary and label encoding would depend on the machine learning algorithm we plan to use. For example, binary encoding may be more suitable for algorithms like logistic regression, while label encoding may be more suitable for decision tree-based algorithms.
2. Education Level: This variable has more than two categories, and the categories have a natural order (i.e., High School < Bachelor's < Master's < PhD). In this case, we can use ordinal encoding, where we assign a numerical value to each category based on its rank. For example, we can assign 0 to High School, 1 to Bachelor's, 2 to Master's, and 3 to PhD. This encoding preserves the order of the categories and allows the machine learning algorithm to capture the relationship between them.
3. Employment Status: This variable also has more than two categories, but the categories do not have a natural order. In this case, we can use one-hot encoding, where we create a binary variable for each category. For example, we can create three binary variables: "Unemployed" (1 if the person is unemployed, 0 otherwise), "Part-Time" (1 if the person is part-time employed, 0 otherwise), and "Full-Time" (1 if the person is full-time employed, 0 otherwise). This encoding allows the machine learning algorithm to treat each category as a separate feature and capture the relationship between the categories.


Binary encoding and label encoding may be suitable for categorical variables with two categories, ordinal encoding may be suitable for categorical variables with ordered categories, and one-hot encoding may be suitable for categorical variables with non-ordered categories.

## Q7. You are analyzing a dataset with two continuous variables, "Temperature" and "Humidity", and two categorical variables, "Weather Condition" (Sunny/Cloudy/Rainy) and "Wind Direction" (North/South/East/West). Calculate the covariance between each pair of variables and interpret the results.

In [17]:
import numpy as np
import pandas as pd
# Set seed for reproducibility
np.random.seed(321)
# Generate data
n = 1000
temp = np.random.normal(25, 5, n)
humidity = np.random.normal(60, 10, n)
weather_condition = np.random.choice(['Sunny', 'Cloudy', 'Rainy'], size=n)
wind_direction = np.random.choice(['North', 'South', 'East', 'West'], size=n)
# Create dataframe
df = pd.DataFrame({
    'Temperature': temp, 
    'Humidity': humidity, 
    'Weather Condition': weather_condition, 
    'Wind Direction': wind_direction
})
# Show first few rows
df.head()

Temperature   Humidity Weather Condition Wind Direction
0    25.862597  50.526311             Sunny          South
1    33.177413  55.809608             Sunny          South
2    25.186682  70.091030             Sunny           West
3    20.579252  68.981094             Sunny          South
4    19.284039  78.624127             Rainy           East

In [18]:
df.cov(numeric_only=True)

Temperature    Humidity
Temperature    25.165416    1.610779
Humidity        1.610779  105.612893

The covariance between "Temperature" and "Humidity" is 1.611 , indicating a positive relationship between the two variables. This means that as temperature increases, humidity tends to increase as well. The variances of each variable are shown on the diagonal, with Humidity having a larger variance than Temperature.

To calculate the covariance between the continuous variables and the categorical variables, we can group the data by the categorical variables and calculate the covariance for each group. Here's an example code:

It is important to note that we cannot calculate the covariance between continuous and categorical variables since covariance requires numerical data. Therefore, we cannot interpret the covariance between "Temperature" and "Weather Condition" or between "Humidity" and "Wind Direction". In general, we need to be careful when interpreting covariance and consider the nature of the variables being analyzed.